swap/swap.go

```diff
@@ -17,69 +17,340 @@
 
-//GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+// Create a Cheque structure emitted to a specific peer as a beneficiary
+// The serial and amount of the cheque will depend on the last cheque and current balance for this peer
+// The cheque will be signed and point to the issuer's contract
+func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
+	var cheque *Cheque
+	var err error
+
+	swapPeer := s.getPeer(peer)
+	beneficiary := swapPeer.beneficiary
+
+	peerBalance := s.balances[peer]
+	// the balance should be negative here, we take the absolute value:
+	honey := uint64(-peerBalance)
+
+	// convert honey to ETH
+	var amount uint64
+	amount, err = s.oracle.GetPrice(honey)
+	if err != nil {
+		log.Error("error getting price from oracle", "err", err)
+		return nil, err
+	}
+
+	// we need to ignore the error check when loading from the StateStore,
+	// as an error might indicate that there is no existing cheque, which
+	// could mean it's the first interaction, which is absolutely valid
+	_ = s.loadCheque(peer)
+	lastCheque := s.cheques[peer]
+
+	if lastCheque == nil {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: uint64(1),
+				Amount: uint64(amount),
+			},
+		}
+	} else {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: lastCheque.Serial + 1,
+				Amount: lastCheque.Amount + uint64(amount),
+			},
+		}
+	}
+	cheque.ChequeParams.Timeout = defaultCashInDelay
+	cheque.ChequeParams.Contract = s.owner.Contract
+	cheque.ChequeParams.Honey = uint64(honey)
+	cheque.Beneficiary = beneficiary
+
+	cheque.Sig, err = s.signContent(cheque)
+
+	return cheque, err
+}
+
+// PeerBalance returns the balance for a given peer
+func (s *Swap) PeerBalance(peer enode.ID) (int64, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	var peerBalance int64
+	var err error
+
+	// load balance from memory
+	peerBalance, keyExists := s.balances[peer]
+	if !keyExists {
+		// if not in memory, try to load balance from disk
+		err = s.stateStore.Get(peer.String(), &peerBalance)
+	}
+	return peerBalance, err
+}
+
+// Balances returns the balances for all known SWAP peers
+func (s *Swap) Balances() (map[enode.ID]int64, error) {
+	balances := make(map[enode.ID]int64)
+
+	// get list of all known SWAP peers
+	swapPeers, err := s.Peers()
+	if err != nil {
+		return nil, err
+	}
+
+	// get balance for list of peers
+	for _, peer := range swapPeers {
+		peerBalance, err := s.PeerBalance(peer)
+		if err != nil {
+			return nil, err
+		}
+		balances[peer] = peerBalance
+	}
+
+	return balances, nil
+}
+
+// Peers returns a list of every peer known through SWAP
+func (s *Swap) Peers() (peers []enode.ID, err error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	knownPeers := make(map[enode.ID]bool)
+
+	// get peer IDs from store
+	storePeers, err := s.stateStore.Keys()
```

Comment:
w00000t??
you realise there's a whole bunch of other stuff in that store right? please remove this function and lets find some other way to deal with this

Reply:
This shouldn't apply anymore now that we have the `prefix` param in the `Keys` function.

Reply:
This anyway is not what @acud though it was, the `stateStore` here is an independent instance and not the global `stateStore` used elsewhere in the code. It only stores balances and cheques. In any case it has been addressed.